swift/extractor/SwiftOutputRewrite.cpp

```diff
@@ -0,0 +1,329 @@
+    newMap.copyFrom(*oldMap);
+    for (auto& entry : newMap) {
+      auto oldPath = entry.getSecond();
+      auto newPath = config.tempArtifactDir + '/' + oldPath;
+      entry.getSecond() = newPath;
+      remapping[oldPath] = newPath;
+    }
+  }
+  std::error_code ec;
+  llvm::SmallString<PATH_MAX> filepath(newPath);
+  llvm::StringRef parent = llvm::sys::path::parent_path(filepath);
+  if (std::error_code ec = llvm::sys::fs::create_directories(parent)) {
+    std::cerr << "Cannot create relocated output map dir: '" << parent.str()
+              << "': " << ec.message() << "\n";
+    return std::nullopt;
+  }
+
+  llvm::raw_fd_ostream fd(newPath, ec, llvm::sys::fs::OF_None);
+  newOutputMap.write(fd, keys);
+  return newPath;
+}
+
+// This is an Xcode-specific workaround to produce alias names for an existing .swiftmodule file.
+// In the case of Xcode, it calls the Swift compiler and asks it to produce a Swift module.
+// Once it's done, Xcode moves the .swiftmodule file in another location, and the location is
+// rather arbitrary. Here are examples of such locations:
+// Original file produced by the frontend:
+//  DerivedData/<Project>/Build/Intermediates.noindex/<Project>.build/<BuiltType>-<Target>/<Project>.build/Objects-normal/<Arch>/<ModuleName>.swiftmodule
+// where:
+//  Project: name of a project, target, or scheme
+//  BuildType: Debug, Release, etc.
+//  Target: macOS, iphoneos, appletvsimulator, etc.
+//  Arch: arm64, x86_64, etc.
+//
+// So far we observed that Xcode can move the module into different locations, and it's not
+// entirely clear how to deduce the destination from the context available for the extractor.
+// 1. First case:
+//  DerivedData/<Project>/Build/Products/<BuiltType>-<Target>/<ModuleName>.swiftmodule/<Arch>.swiftmodule
+//  DerivedData/<Project>/Build/Products/<BuiltType>-<Target>/<ModuleName>.swiftmodule/<Triple>.swiftmodule
+// 2. Second case:
+//  DerivedData/<Project>/Build/Products/<BuiltType>-<Target>/<ModuleName>/<ModuleName>.swiftmodule/<Arch>.swiftmodule
+//  DerivedData/<Project>/Build/Products/<BuiltType>-<Target>/<ModuleName>/<ModuleName>.swiftmodule/<Triple>.swiftmodule
+// 2. Third case:
+//  DerivedData/<Project>/Build/Products/<BuiltType>-<Target>/<ModuleName>/<ModuleName>.framework/Modules/<ModuleName>.swiftmodule/<Arch>.swiftmodule
+//  DerivedData/<Project>/Build/Products/<BuiltType>-<Target>/<ModuleName>/<ModuleName>.framework/Modules/<ModuleName>.swiftmodule/<Triple>.swiftmodule
+// The <Triple> here is a normalized target triple (e.g. arm64-apple-iphoneos15.4 ->
+// arm64-apple-iphoneos).
+//
+// This method constructs those aliases for a module only if it comes from Xcode, which is detected
+// by the presence of an `Intermediates.noindex` directory in the module path.
+//
+// In the case of the Swift Package Manager (`swift build`) this is not needed.
+static std::vector<std::string> computeModuleAliases(llvm::StringRef modulePath,
```

Comment:
this function is quite a lot to chew through... I think it's ok for now, but maybe I'd leave a todo, so that by the next time we need to do something here, if we finally have access to `std::filesystem` we should definitely use that instead of all the string operations here.